### DISCUSSION TITLE
Is this supposed to be catched by reportIncompatibleMethodOverride ?

### DISCUSSION THREAD

**Discussion by mehdigmira** [microsoft/pyright#4134] [Q&A]
The example below, will raise at runtime, but is not catched by reportIncompatibleMethodOverride.
This seems to be s too simple example for this to be a bug. So what is the rational behind allowing this pattern ?

```
class A:
    def f(self, x: int):
        ...


class B(A):
    def f(self, x: int, *args: int):
        super().f(x, *args)


B().f(1, 2, 3)
```

Also i'm a bit uncertain of the exact usecases that reportIncompatibleMethodOverride catches, especially when using *args **kwargs and  default keyword arguments

**EDIT**

Actually the issue is maybe not with reportIncompatibleMethodOverride, but with super() call not being flagged as an error.
I don't understand why

```
# pyright: strict


def g(x: int):
    ...


def f(x: int, *args: int):
    g(x, *args)
```

is allowed even in strict mode. What kind of pattern are we trying to support here ?

**Comment by erictraut** [comment]
This is related to issue https://github.com/microsoft/pyright/issues/4118. The `reportIncompatibleMethodOverride` is currently not as strict as it could be. It reports errors for common coding mistakes but does not report errors for common coding patterns in Python that are potentially safe. One common pattern is to override a base method with a `(*args, **kwargs)` method.

The reason https://github.com/microsoft/pyright/issues/4118 is marked as an enhancement request rather than a bug is that the current implementation is working "as designed", but I think it's reasonable to augment it to be stricter and report more overrides as mismatches. I'll need to think more about which cases are appropriate to handle.

Our general philosophy in pyright is that false positive errors are worse than false negatives. When we need to make a tradeoff between the two, we err on the side of avoiding false positives.

**Reply by mehdigmira** [reply]
I think a stricter reportIncompatibleMethodOverride could be very useful indeed. And i do find pyright's philosophy to be the correct one.

However the issue is not related to the reportIncompatibleMethodOverride flag.

```
def g(x: int):
    ...


g(1, *(2, 3))  # no error here
```

How come this code is valid ? Did i miss smth ?

**Reply by erictraut** [reply]
Yes, that's a false negative. I've filed https://github.com/microsoft/pyright/issues/4140 to track this.

**Comment by mehdigmira** [comment]
To continue on `reportIncompatibleMethodOverride`, I find it a bit hard to understand

```
T = TypeVar("T")


class A:
    def f(self: T) -> list[T]:
        ...


class B(A):
    def f(self) -> list["B"]:  # pyright reports an error here
        return [B()]
```

for example, here an error is reported where it should not IMO. Switching from list to sequence makes this work. But i don't see why we would need a covariant type here.
